### PR TITLE
fix: `bitSequence`

### PR DESCRIPTION
**Pull request by josepot**
I completely misunderstood how the `bitSequence` type of the metadata works. Big kudos to @kratico for helping me debug the issue! 🙏

### CODE REVIEW

**Comment by ryanleecode** [comment]
Looks good. Lets also add a test for the codec.

File: packages/substrate-codecs/src/bitSequence.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { Bytes, Decoder, Encoder, createCodec, createDecoder } from "scale-ts"
+import { compactNumber } from "./compact"
+
+export interface BitSquence {
```
Comment (ryanleecode) [inline packages/substrate-codecs/src/bitSequence.ts:4]: BitSequence

**Comment by ryanleecode** [comment]
What would be the expected behaviour for trying to decode an input than is less than `bitSize`? i.e lets say bitSize is equivalent to 10 bytes but the "otherData" is 5 bytes.

**Comment by josepot** [comment]
> What would be the expected behaviour for trying to decode an input than is less than `bitSize`? i.e lets say bitSize is equivalent to 10 bytes but the "otherData" is 5 bytes.

The same one as trying to decode anything that's expected to have `n` bytes, but that in reality it has less than `n` bytes... You could ask the same question about any other decoder, really. The expected behaviour is a runtime error.

File: packages/substrate-codecs/src/bitSequence.ts (hunk 1 above)
Comment (josepot) [reply packages/substrate-codecs/src/bitSequence.ts]: Thanks! Fixed!

**Comment by josepot** [comment]
> What would be the expected behaviour for trying to decode an input than is less than `bitSize`? i.e lets say bitSize is equivalent to 10 bytes but the "otherData" is 5 bytes.
> 
> this is an edge case since it would assume its is the last element in the entire encoded data

I guess that maybe you meant that the encoder should throw when it receives more bits than the possible available bytes in the same input? 🤔

I mean -in general terms- this library (and `scale-ts`) is very unapologetic about expecting the inputs to be validated by the consumer (when they must be validated, which BTW is not always the case). However, I guess that someone could argue that this could be a worthy exception. So, I added a new commit to handle this edge-case.

**Comment by ryanleecode** [comment]
Nope i mean for decoding. 

> The same one as trying to decode anything that's expected to have n bytes, but that in reality it has less than n bytes... You could ask the same question about any other decoder, really (e.g u16.dec('0x01')). The expected behaviour is a runtime error.

So if this is true then this should throw right

```ts
bitSequence.dec(concatBytes(compact.enc(16), randomBytes(1)))
```

because this returns

```ts
{ bytes: Uint8Array(1) [ 44 ], bitsLen: 16 }
```

File: packages/substrate-codecs/src/bitSequence.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  const bitsLen = compactNumber.dec(data)
+  const bytesLen = Math.ceil(bitsLen / 8)
+  const bytes = Bytes(bytesLen).dec(data)
+  return { bytes, bitsLen }
+})
+
+const bitSequenceEncoder: Encoder<BitSequence> = (input) => {
+  if (input.bitsLen > input.bytes.length * 8)
+    throw new Error(
+      `Not enough bits. (bitsLen:${input.bitsLen}, bytesLen:${input.bytes.length})`,
```
Comment (ryanleecode) [inline packages/substrate-codecs/src/bitSequence.ts:19]:
```suggestion
      `Not enough bytes. (bitsLen:${input.bitsLen}, bytesLen:${input.bytes.length})`,
```

Comment (josepot) [reply packages/substrate-codecs/src/bitSequence.ts:19]: Nope. The error is correct... there are not enough `bits` inside the provided `bytes` 🙂

Comment (ryanleecode) [reply packages/substrate-codecs/src/bitSequence.ts:19]: wouldnt it be confusing because they would think you're referring to `bitsLens`

**Comment by josepot** [comment]
> Nope i mean for decoding.
> 
> > The same one as trying to decode anything that's expected to have n bytes, but that in reality it has less than n bytes... You could ask the same question about any other decoder, really (e.g u16.dec('0x01')). The expected behaviour is a runtime error.
> 
> So if this is true then this should throw right
> 
> ```ts
> bitSequence.dec(concatBytes(compact.enc(16), randomBytes(1)))
> ```
> 
> because this returns
> 
> ```ts
> { bytes: Uint8Array(1) [ 44 ], bitsLen: 16 }
> ```

Aha! I see! Ok, so the only thing that I know for sure is that the behaviour should be consistent with the behaviour of the `Bytes` decoder. So, this is the same as asking:

> what should be the expected behaviour for `Bytes(2).dec('0x01')`?

And to me the correct answer is either: "who cares?" or "why does it matter?".

If you feel strongly that it's important to have guarantees and expected behaviours for wrong inputs **on decoder functions**, then please let me know when this would be relevant for the problems that we are trying to solve with this library 🙏.

I'm of the opinion that it's completely fine for these kinds of decoders to only provide guarantees and expected behaviours for when they are used with correct inputs. Please let's keep in mind that the inputs of these kinds of decoders come directly from on-chain binary data, which should always be compliant with the created decoders.

**Comment by ryanleecode** [comment]
Yah thats fair, the only way it can happen is if there is a failure on the encoding side (from say a dapp), but we already ruled that out by throwing an error so thats not an issue. 

LGTM then

**Review by ryanleecode — APPROVED** [review]

**Comment by josepot** [comment]
> Yah thats fair, the only way it can happen is if there is a failure on the encoding side (from say a dapp), but we already ruled that out by throwing an error so thats not an issue.

It's honestly hard for me to picture a dApp where users are asked to input a `bitSequence` 😅

However, jokes aside, diving deeper into what I mentioned earlier:
1. Our main goal is to ensure that our generated decoders, which use metadata to read on-chain data, work flawlessly. If we succeed here, we can avoid potential pitfalls.
2. But, we have to be pragmatic. No matter how thorough our tests are, issues will inevitably happen. When they do, I think it's best to integrate error handlers and tracing mechanisms specifically in the more complex decoders that we have composed (using the metadata) to interpret on-chain data.

Take the `bitSequence` issue as an example: We discovered the error when decoding the body entries of the latest blocks. In cases like these, it's beneficial for the `call` decoder to identify the error, log the binary data it was trying to decode, and pinpoint where within its nested decoders the mistake happened. IMO: this level of granularity in error-handling simplifies debugging and can helps us "quickly" understand what the problem was. However, I'm skeptical about the utility of managing exceptions in our basic, low-level decoders.

**Pull request merged**